src/src_main/crypto_tc.c

```diff
@@ -407,195 +407,66 @@ int32_t Crypto_TC_ApplySecurity(const uint8_t *p_in_frame, const uint16_t in_fra
+            }
 
-                mac_loc = TC_FRAME_HEADER_SIZE + segment_hdr_len + SPI_LEN + sa_ptr->shivf_len + sa_ptr->shsnf_len +
-                          sa_ptr->shplf_len + tf_payload_len;
-#ifdef MAC_DEBUG
-                printf(KYEL "MAC location is: %d\n" RESET, mac_loc);
-                printf(KYEL "MAC size is: %d\n" RESET, MAC_SIZE);
+#ifdef TC_DEBUG
+            printf("Encrypted bytes output_loc is %d\n", index);
+            printf("Input bytes input_loc is %d\n", TC_FRAME_HEADER_SIZE + segment_hdr_len);
 #endif
-                gcry_error =
-                    gcry_cipher_gettag(tmp_hd,
-                                       &p_new_enc_frame[mac_loc], // tag output
-                                       MAC_SIZE // tag size // TODO - use sa_ptr->abm_len instead of hardcoded mac size?
-                    );
-                if ((gcry_error & GPG_ERR_CODE_MASK) != GPG_ERR_NO_ERROR)
-                {
-                    printf(KRED "ERROR: gcry_cipher_checktag error code %d\n" RESET, gcry_error & GPG_ERR_CODE_MASK);
-                    status = CRYPTO_LIB_ERR_MAC_RETRIEVAL_ERROR;
-                    return status;
-                }
-                // Zeroise any sensitive information
-                gcry_cipher_close(tmp_hd);
+
+            if(ecs_is_aead_algorithm == CRYPTO_TRUE)
+            {
+                status = cryptography_if->cryptography_aead_encrypt(&p_new_enc_frame[index],                               // ciphertext output
+                                                           tf_payload_len,                                        // length of data
+                                                           (uint8_t*)(p_in_frame + TC_FRAME_HEADER_SIZE + segment_hdr_len), // plaintext input
+                                                           tf_payload_len,                                         // in data length
+                                                           NULL, // Using SA key reference, key is null
+                                                           KEY_SIZE, // Length of key. TODO - why is this hard-coded?
+                                                           sa_ptr, // SA (for key reference)
+                                                           sa_ptr->iv, // IV
+                                                           sa_ptr->shivf_len, // IV Length
+                                                           mac_ptr, // tag output
+                                                           MAC_SIZE, // tag size // TODO - why is this hard-coded?!
+                                                           aad, // AAD Input
+                                                           aad_len, // Length of AAD
+                                                           (sa_ptr->est==1),
+                                                           (sa_ptr->ast==1),
+                                                           (sa_ptr->ast==1)
```

Comment:
Remove duplicated pointer

Reply:
Should be (sa_ptr->aad==1) ?  With aad instead of ast?

Reply:
I think I might agree with you, these three parameters refer to the following booleans:
                                         uint8_t decrypt_bool, uint8_t authenticate_bool,
                                         uint8_t aad_bool)

The question is, will you ever provide AAD, but not authenticate? Or Authenticate, but not provide AAD? If the answer is no to both questions, then we can combine those both into a single authenticate_bool.

For some reason during my refactor, I thought I ran into one such case:
https://github.com/nasa/CryptoLib/pull/46/files#diff-680e5a3e45c21aa3a202f9dcff50de85c32aad656eaae798a6052c7495c7df59R106

But that may be incorrect (there was also the issue with it doing authenticate then encrypt, instead of encrypt then authenticate). I will investigate some more.

Reply:
For now, I think it's OK to keep the interface function as having different controls for authentication vs AAD related calls... In the case of TC, the presence of AAD is determined by the AST flag in the SA... I don't know about the TM & key management cases, as linked above, it seems to vary, but that logic may be wrong. For the purposes of this initial interface implementation, it should be OK to keep it separate, as it is now.